### PR TITLE
MLX: bring DPO's reference policy to parity with TRL's DPOTrainer

### PR DESCRIPTION
**Pull request by Lyxot**
Builds on #1164 (merged); the diff is against `main`.

## What changes

MLX DPO's frozen reference policy now covers what TRL 0.23.1's `DPOTrainer` covers, at the cost TRL pays per step, and exposes the same reference configuration: `ref_model` as a constructor argument plus `model_adapter_name`, `ref_adapter_name`, `force_use_ref_model`, `sync_ref_model`, `ref_model_mixup_alpha`, `ref_model_sync_steps`, `precompute_ref_log_probs` and `precompute_ref_batch_size` on `MLXDPOConfig`. Four commits, one per capability:

1. **A reference through adapter overrides, DoRA, and a saved adapter.** The reference is a `ReferencePolicy`: a scale per adapter module plus an array per parameter, applied to the policy model while a loss forward, a compiled step or a reference decode runs and restored afterwards, captured in every compile state list so a compiled step reads it through the same tree as the model's own. It covers the base model with the adapter disabled (LoRA scale zero; DoRA's magnitude recomputed from the base weight, so a trained `m` needs no persisted copy), trainable tensors no adapter owns held at their starting value (as PEFT restores `modules_to_save`), and `ref_adapter_name` naming a saved adapter directory whose tensors and scale stand in for the model's. `model_adapter_name` accepts only the single unnamed adapter, as `"default"`.
2. **A `ref_model`, or a full fine-tune's starting weights.** `build_reference_policy` picks the reference in TRL's order: a `ref_model` of its own, else the model with its adapters disabled, else the model at the weights it starts from. A `ref_model` is put in eval mode, its norm weights kept in float32 as the policy's are, and its state captured by the compiled step. A full fine-tune pins every trainable tensor at its starting array as an override on the module that owns it; arrays are immutable, so this copies nothing until the optimizer replaces a tensor. Provenance kinds `reference_model` and `parameter_snapshot` check a resume against the reference it started with.
3. **TR-DPO (`sync_ref_model`).** Every `ref_model_sync_steps` optimizer steps the reference becomes `(1 - alpha) * reference + alpha * policy`, mixed in float32 and stored at the reference's dtype, over the policy's trainable tensors by name. A snapshot or a fresh adapter (LoRA or DoRA) syncs in place inside the list a compiled step captured, so the next step reads the mix without a retrace; a `ref_model` must own every trainable path at the same shape. Each checkpoint of a synced run carries `reference.safetensors`, and a resumed run restores it.
4. **Precomputed reference log probabilities.** `precompute_ref_log_probs` scores the training rows before the first step and each eval split when its first evaluation builds it, `precompute_ref_batch_size` pairs at a time (the split's batch size when unset). A batch then carries its rows' values as a fourth element, and the loss and evaluation read them instead of running the reference, so a step costs one policy forward. Once nothing reads it, a snapshot reference drops its arrays, so full-fine-tuning DPO runs at one model's memory.

## Behavior changes and TRL divergences reviewers should know

- **Compatibility break:** a model reloaded from a saved adapter that also carries a trained norm used to train against a silent hybrid (base weights plus the trained norm); it is now refused, with `ref_adapter_name` or `reference_free=True` named as the ways forward. The loader records the restored non-adapter keys on the model for this check.
- `ref_model` on an adapted model warns (the base is already the reference and a second model doubles memory) where TRL refuses; `force_use_ref_model=True` silences it. `ref_model` together with `ref_adapter_name` is refused rather than ignoring the adapter name, and `ref_model` may not be the policy itself.
- TR-DPO extensions over TRL: `ref_model=None` is accepted (the snapshot syncs), a trained adapter is refused since the base cannot own its delta, `ref_model_mixup_alpha` must lie in `[0, 1]` (TRL's own convex-mix definition), frozen tensors of a `ref_model` never drift, and the synced reference is persisted per checkpoint, which TRL does not do (a resumed TR-DPO run would otherwise restart its reference at the base).
- Precompute: a row's precomputed value is what the live forward gives it in a batch of that shape. An eval split's chunks are its batches by default, so its values are the live ones exactly; a training batch is shaped by its schedule, so its values agree to the kernels' rounding at another shape (on a 4-bit 135M model a row's log probability moves by up to a few tenths of a nat between one and two pairs; TRL's sequential precompute dataloader has the same property against its shuffled training batches). Refused with `sync_ref_model`; ignored by a reference-free objective.

## Validation

Unit files, under the `mlx_simulation` shim and on Metal:

```bash
python -m pytest tests/test_mlx_preference.py tests/test_mlx_preference_reference_metal.py -q
```

The compiled-step test runs each reference form (adapter, full fine-tune, `ref_model`, and every synced form, DoRA and an adapted `ref_model` included) inside an `mx.compile`d step against a twin model as the oracle. The precompute lifecycle test pins the chunking, that the eval split is scored with training patches paused and the model in eval mode, the snapshot's release, and the refusals.

During development each commit was also checked against a broader set kept as development validation rather than repository tests: mutation checks of every production change (each mutation fails a named test; 30 to 37 per stage), end-to-end LoRA, DoRA and full-fine-tune DPO runs on a 4-bit 135M model with `mx.compile`, TR-DPO checkpoint and resume runs whose saved reference equals `0.75 * base + 0.25 * policy` at the step it was taken, and a precomputed run whose eval values are bitwise those the live run scores at its evaluation.

### CODE REVIEW

**Comment by danielhanchen** [comment]
This also refuses runs that reload a saved adapter carrying a trained norm, which used to train against the hybrid reference, so anyone on that setup would need ref_adapter_name or reference_free=True after upgrading. Could you confirm how you would like that surfaced in a release note before I review, and ping me once #1164 lands and this comes out of draft?

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/preference.py [hunk 1]
```diff
@@ -1540,51 +1705,339 @@ def has_values(value):
...
+                        "tensor of the model into the same-named tensor of "
+                        f"ref_model, which has no floating {name!r} of shape "
+                        f"{tuple(value.shape)}. Load the reference the way the "
+                        "model is loaded, adapters included."
+                    )
+                mirrored.append(_owner_of(ref_by_name, name, of="ref_model"))
+        provenance.update(
+            kind="reference_model",
+            reference_repo=getattr(ref_model, "_hf_repo", None),
+            reference_commit=getattr(ref_model, "_unsloth_base_commit_hash", None),
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/preference.py:1912]: Not changing this here, to keep the PR to the reference forms themselves.

The provenance check is a guard TRL's `DPOTrainer` does not have at all: it resumes with whatever `ref_model` the caller passes and never compares it to the one the checkpoint was written under. This check catches what the checkpoint can record cheaply (a different base repository or commit, a different adapter layout, a synced reference resumed as frozen or the reverse). Telling apart two adapted or locally modified `ref_model`s built from the same base would need a hash over every parameter at each build, or a copy of the reference in every checkpoint, which doubles checkpoint size for a run whose reference never changes. As in TRL, passing the same `ref_model` on resume stays the caller's responsibility. The case where the reference does change during the run, `sync_ref_model`, is the one persisted per checkpoint (`reference.safetensors`) and restored on resume.

File: unsloth_zoo/mlx/preference.py [hunk 2]
```diff
@@ -1540,51 +1705,339 @@ def has_values(value):
...
+            f"Unsloth MLX DPO: ref_adapter_name={str(path)!r} is not a saved "
+            "adapter directory: it has no adapters.safetensors."
         )
-    if any(name not in adapters for name in trainable):
+    tensors = mx.load(str(weights_file))
+    config_file = path / "adapter_config.json"
+    config = {}
+    if config_file.is_file():
+        with open(config_file, encoding="utf-8") as handle:
+            config = json.load(handle)
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/preference.py:1759]: Not changing this here.

`_reference_adapter_overrides` reads only `fine_tune_type` and the LoRA scale from `adapter_config.json`; it does not read a base repository, revision, commit, or quantization, so nothing is read and then ignored. What it checks is what makes the override well-defined: the adapter kind, every adapter tensor of the model present at the model's shape, and the model's trained tensors covered.

Refusing an adapter whose recorded `base_model_name_or_path` differs from the live model's would be a check neither TRL nor PEFT makes (`ref_adapter_name` in TRL is a `set_adapter` call, and PEFT's `load_adapter` fills the field in when it is missing rather than comparing it), and it would reject legitimate setups: the same base loaded from a local directory or under another repository id, or an adapter reused against a different quantization of the same weights. A checkpoint written by this trainer does record `base_model_name_or_path` and the base quantization in its `adapter_config.json`, so a caller can check the pairing before passing it; adding that refusal is out of this PR's scope.

**Comment by Lyxot** [comment]
@codex review

**Comment by danielhanchen** [comment]
This still turns the saved adapter plus trained norm case into a hard error rather than the old hybrid reference, so existing runs would stop after upgrading; could you confirm the wording you want for that in the release note, since my earlier question on it is still open?

**Comment by danielhanchen** [comment]
#1164 has landed and this is out of draft, so the one thing still open on my side is the new refusal for a reloaded adapter carrying a trained norm, which stops a setup that trains today. Could you confirm the release note wording you want for it before I review?

**Comment by danielhanchen** [comment]
Checked this against main, where the reference is still only the LoRA-disabled form in `unsloth_zoo/mlx/preference.py`, so the `ref_model`, TR-DPO and precompute paths are all new here.

Two things before I review: could you say how you want the new refusal of a reloaded adapter carrying a trained norm worded for the release notes, and rebase since this is conflicting now?

**Comment by Lyxot** [comment]
@codex review
